Add the AppendDiscoveryResult to NdrRegistry

DuncanHopkinsFoundry opened the pull request:
This method allows the client to explicitly set additional discovery
results that would NOT be found through the plugin system. For example
to support lazily-loaded plugins which cannot be easily discovered in
advance.

- [X ] I have submitted a signed Contributor License Agreement

diegojesus-foundry committed:
Add the AppendDiscoveryResult to NdrRegistry

This method allows the client to explicitly set additional discovery
results that would NOT be found through the plugin system. For example
to support lazily-loaded plugins which cannot be easily discovered in
advance.

jilliene commented:
Filed as internal issue #USD-7288

spiffmon commented on pxr/usd/ndr/registry.h:
```diff
@@ -86,6 +86,21 @@ class NdrRegistry : public TfWeakBase
+    /// This method will not immediately spawn a parse call which will be
+    /// deferred until a GetNode*() method is called.
+    NDR_API
+    void AppendDiscoveryResult(NdrNodeDiscoveryResult&& discoveryResult);
```
"Append" implies knowledge that we're storing DR's in a list, which is an implementation detail we'd rather not be locked into.  Can we change this to "Add"?

DuncanHopkinsFoundry replied:
Done

spiffmon commented on pxr/usd/ndr/registry.h:
```diff
@@ -86,6 +86,21 @@ class NdrRegistry : public TfWeakBase
+    NDR_API
+    void AppendDiscoveryResult(NdrNodeDiscoveryResult&& discoveryResult);
+
+    /// Copy version of the method above.
```
Can we use the doxygen `\overload` directive, here?

DuncanHopkinsFoundry replied:
Done

spiffmon commented on pxr/usd/ndr/registry.h:
```diff
@@ -86,6 +86,21 @@ class NdrRegistry : public TfWeakBase
+    void AppendDiscoveryResult(NdrNodeDiscoveryResult&& discoveryResult);
+
+    /// Copy version of the method above.
+    /// For performance reasons, one should prefer to use the move version.
```
Can we say "prefer to use the rvalue reference form"?

DuncanHopkinsFoundry replied:
Done

diegojesus-foundry committed:
Rename AppendDiscoveryResult methods to AddDiscoveryResult

Append implies that discovery results are stored in a list while add is
more generic and does not expose implementation details.

This commit also improves these methods' documentation.
